src/Voron/Data/BTrees/TreeMutableState.cs

```diff
@@ -1,39 +1,84 @@
 ﻿using System;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 using Voron.Global;
 using Voron.Impl;
 
 namespace Voron.Data.BTrees
 {
-    public sealed unsafe class TreeMutableState
+
+    [StructLayout(LayoutKind.Explicit, Size=64)]
+    public unsafe struct TreeMutableState(LowLevelTransaction tx)
     {
-        private readonly LowLevelTransaction _tx;
+        // The idea behind the tree mutable state to be a struct is two-fold. First and foremost,
+        // to avoid the allocation of yet another object that serves no other purpose than to be
+        // associated for the lifetime of the tree. Second, to make the access of those fields
+        // and the serialization to be as efficient as possible. For that reason the structure
+        // of the mutable state will mimic the structure that it is written to disk.
 
+        [FieldOffset(0)]
         public RootObjectType RootObjectType;
-        public long RootPageNumber;
-        public TreeFlags Flags;
-
-        private bool _isModified;
 
+        [FieldOffset(1)]
+        public TreeFlags Flags;
+        [FieldOffset(2)]
+        public long RootPageNumber;
+        [FieldOffset(10)]
         public long BranchPages;
+        [FieldOffset(18)]
         public long LeafPages;
+        [FieldOffset(34)]
         public long OverflowPages;
-        public int Depth;
+        [FieldOffset(42)]
         public long PageCount;
+        [FieldOffset(50)]
         public long NumberOfEntries;
+        [FieldOffset(58)]
+        public int Depth;
+
+        [FieldOffset(62)]
+        private bool _isModified;
+        [FieldOffset(63)]
+        public readonly bool IsReadOnly = tx.Flags != TransactionFlags.ReadWrite;
+
 
-        public TreeMutableState(LowLevelTransaction tx)
+        static TreeMutableState()
         {
-            _tx = tx;
+#if DEBUG
+            // We will check in debug builds if the structures are compatible. Any change to the TreeMutableState
+            // or the TreeRootHeader must be reflected on both. 
+            var state = new TreeMutableState
+            {
+                RootObjectType = RootObjectType.VariableSizeTree, 
+                Flags = TreeFlags.FixedSizeTrees, 
+                RootPageNumber = 32, 
+                BranchPages = 64,
+                LeafPages = 128,
+                OverflowPages = 256,
+                PageCount = 512,
+                NumberOfEntries = 1024,
+                Depth = 2048,
+                IsModified = true
+            };
+
+            ref var rootHeader = ref Unsafe.As<TreeMutableState, TreeRootHeader>(ref state);
```

Comment:
You are trying to get cast identical, but why?
Also, if they are structurally the same, why wouldn't this just be `TreeRootHeader` ?

Reply:
They are not structurally same, just structurally compatible. The idea behind the sequence of commits that will comprise this experimental PR is to optimize the hell out of the sections that happen under the lock (even if at the expense of maintainability). Then we can mix and match which ones we are OK to keep, and push them as individual optimizations. For example, in this case I got a very modest 500MB of allocations less (in 36Gb is not that much), but it proved that pushing this type of optimizations are easier (even if high riks) than what I would have expected.

Reply:
What's the difference between those?
Why not just use `TreeRootHeader` directly?

Reply:
Can't we add that to the `TreeRootHeader` directly, then just change the type? Or methods